src/wing/CCPACSWing.cpp

```diff
@@ -420,49 +425,68 @@ void CCPACSWing::BuildUpperLowerShells()
-
-       const CCPACSControlSurfaces& controlSurfs = componentSegment.GetControlSurfaces().value();
-       if (!controlSurfs.GetTrailingEdgeDevices().is_initialized()) continue;
-       const CCPACSTrailingEdgeDevices& controlSurfaceDevices = controlSurfs.GetTrailingEdgeDevices().value();
+    for (int i = 1; i <= GetComponentSegmentCount(); i++) {
 
-        for ( size_t j = controlSurfaceDevices.GetTrailingEdgeDevices().size(); j > 0 ; j-- ) {
-            CCPACSTrailingEdgeDevice& controlSurfaceDevice = *controlSurfaceDevices.GetTrailingEdgeDevices().at(j-1);
+        const CCPACSWingComponentSegment& componentSegment = GetComponentSegment(i);
+        if (!componentSegment.GetControlSurfaces().is_initialized()) {
+            continue;
+        }
 
-            PNamedShape controlSurfacePrism = controlSurfaceDevice.GetCutOutShape();
-            if (controlSurfaceDevice.GetType() != SPOILER) {
-                if (!first) {
-                    ListPNamedShape childs;
-                    childs.push_back(controlSurfacePrism);
-                    fusedBoxes = CFuseShapes(fusedBoxes, childs);
-                }
-                else {
-                    first = false;
-                    fusedBoxes = controlSurfacePrism;
+        const CCPACSControlSurfaces& controlSurfs = componentSegment.GetControlSurfaces().value();
+        if (controlSurfs.GetTrailingEdgeDevices().is_initialized()) {
+            const CCPACSTrailingEdgeDevices& TrailingEdgeDevices = controlSurfs.GetTrailingEdgeDevices().value();
+
+            for (size_t ted_index = 0; ted_index < TrailingEdgeDevices.GetTrailingEdgeDevices().size(); ted_index++) {
+                CCPACSTrailingEdgeDevice& TrailingEdgeDevice = *TrailingEdgeDevices.GetTrailingEdgeDevices().at(ted_index);
+
+                PNamedShape TrailingEdgeDevicePrism = TrailingEdgeDevice.GetCutOutShape();
+                    if (!first) {
+                        ListPNamedShape childs;
+                        childs.push_back(TrailingEdgeDevicePrism);
+                        fusedBoxes = CFuseShapes(fusedBoxes, childs);
+                    }
+                    else {
+                        first      = false;
+                        fusedBoxes = TrailingEdgeDevicePrism;
+                    }
+                // trigger build of the flap
+                TrailingEdgeDevice.GetLoft();
+            }
+        }
+        if (controlSurfs.GetLeadingEdgeDevices().is_initialized()) {
+            const CCPACSLeadingEdgeDevices& LeadingEdgeDevices = controlSurfs.GetLeadingEdgeDevices().value();
+
+            for (size_t led_index = 0; led_index < LeadingEdgeDevices.GetLeadingEdgeDevices().size(); led_index++) {
+                CCPACSLeadingEdgeDevice& LeadingEdgeDevice = *LeadingEdgeDevices.GetLeadingEdgeDevices().at(led_index);
+
+                PNamedShape LeadingEdgeDevicePrism = LeadingEdgeDevice.GetCutOutShape();
+                if (LeadingEdgeDevice.GetType() != SPOILER) {
```

Comment:
I missed this line in my previous review. A LeadingEdgeDevice never is a spoiler.